src/main/java/hudson/plugins/ec2/EC2AbstractSlave.java

```diff
@@ -485,17 +508,26 @@ void stop() {
 
     }
 
-    boolean terminateInstance() {
-        try {
+    private boolean isUsingDynamicSshKeys() {
+        EC2Cloud cloud = getCloud();
+        if (cloud != null) {
+            String id = cloud.getSshKeysCredentialsId();
+            if ((id == null) || (id.isEmpty())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    protected void cleanupSshKeyPairIfNeeded() {
+        if (getInstanceSshKeyPairName() != null) {
+            //this instance is using dynamic ssh keys
+            LOGGER.fine(() -> "EC2 instance delete key pair request sent for " + this.getInstanceSshKeyPairName());
             AmazonEC2 ec2 = getCloud().connect();
-            TerminateInstancesRequest request = new TerminateInstancesRequest(Collections.singletonList(getInstanceId()));
-            LOGGER.fine("Sending terminate request for " + getInstanceId());
-            ec2.terminateInstances(request);
-            LOGGER.info("EC2 instance terminate request sent for " + getInstanceId());
-            return true;
-        } catch (AmazonClientException e) {
-            LOGGER.log(Level.WARNING, "Failed to terminate EC2 instance: " + getInstanceId(), e);
-            return false;
+            ec2.deleteKeyPair(new DeleteKeyPairRequest().withKeyName(this.getInstanceSshKeyPairName()));
+        } else {
+            // should not be possible to reach this code
+            LOGGER.warning(() -> "Unable to determine keypair name to delete for this instance");
```

Comment:
No longer true IIUC; rather, the normal case when deleting an agent using the original manually configured keypair?

Anyway you need to note the name of the agent or the log message is useless:

```suggestion
            LOGGER.fine(() -> "No dynamic keypair to delete for " + getNodeName());
```